clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/LeaderFollowerStoreIngestionTask.java

```diff
@@ -1895,6 +1900,169 @@ private void recordRegionHybridConsumptionStats(
+
+    if (isLagAcceptable && isHybridFollower(pcs)) {
+      // non AA stores have issues reading HB SOS from the RT leading to the standby replicas waiting for
+      // leader completion state header indefinitely, so disabling it until that issue is resolved
+      if (!(getServerConfig().isLeaderCompleteStateCheckEnabled() && this.isActiveActiveReplicationEnabled())) {
+        isLagAcceptable = true;
+      } else if (!pcs.isFirstHeartBeatSOSReceived()) {
+        // wait for the first HB to know if the leader supports sending LeaderCompleteState or not
+        isLagAcceptable = false;
+      } else if (pcs.getLeaderCompleteState().equals(LEADER_COMPLETE_STATE_UNKNOWN)) {
+        // if the leader don't support LeaderCompleteState
+        isLagAcceptable = true;
+      } else {
+        // check if the leader is completed and the last update time was within the configured time
+        isLagAcceptable = pcs.isLeaderCompleted()
+            && ((System.currentTimeMillis() - pcs.getLastLeaderCompleteStateUpdateInMs()) <= getServerConfig()
+                .getLeaderCompleteStateCheckValidIntervalMs());
+      }
+    }
+
+    if (shouldLogLag) {
+      String lagLogFooter;
+      if (isHybridFollower(pcs)) {
+        lagLogFooter = ". Leader Complete State: {" + pcs.getLeaderCompleteState().toString()
+            + "}, Last update In Ms: {" + pcs.getLastLeaderCompleteStateUpdateInMs() + "}.";
+      } else {
+        lagLogFooter = "";
+      }
+      LOGGER.info(
+          "{} [{} lag] partition {} is {}lagging. {}Lag: [{}] {} Threshold [{}]{}",
+          isOffsetBasedLag ? "Offset" : "Time",
+          consumerTaskId,
+          pcs.getPartition(),
+          (isLagAcceptable ? "not " : ""),
+          (isOffsetBasedLag ? "" : "The latest producer timestamp is " + latestConsumedProducerTimestamp + ". "),
+          offsetLag,
+          (isLagAcceptable ? "<" : ">"),
+          offsetThreshold,
+          lagLogFooter);
+    }
+
+    return isLagAcceptable;
+  }
+
+  /**
+   * HeartBeat SOS messages carry the leader completion state in the header. This function extracts the leader completion
+   * state from that header and updates the {@param partitionConsumptionState} accordingly. <p>
+   * If there is no leader completion state header, reset the leader completion state to
+   * {@link LeaderCompleteState#LEADER_COMPLETE_STATE_UNKNOWN} as the leader don't know about this header
+   * (using old version of the code) or the leader may have rolled back to a version that doesn't support this header or
+   * this topic partition gets a new leader which doesn't support this header yet.
+   */
+  protected void getAndUpdateLeaderCompletedState(
+      PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long> consumerRecord,
+      PartitionConsumptionState partitionConsumptionState) {
+    if (isHybridFollower(partitionConsumptionState)) {
+      ControlMessage controlMessage = (ControlMessage) consumerRecord.getValue().payloadUnion;
```

Comment:
So the passed message is guaranteed to be `ControlMessage`? Shall we have a type check here?

Reply:
Another way would be passing `ControlMessage` param directly along with other required info, such as headers.

Reply:
It was called from only one place and there was a check before that for control Message. But modified it to pass the CM param along with other params directly rather than the `consumerRecord`.